client/network/src/protocol/generic_proto/handler/group.rs

```diff
@@ -461,21 +568,93 @@ impl ProtocolsHandler for NotifsHandler {
 	) -> Poll<
 		ProtocolsHandlerEvent<Self::OutboundProtocol, Self::OutboundOpenInfo, Self::OutEvent, Self::Error>
 	> {
+		if let Some(notifications_sink_rx) = &mut self.notifications_sink_rx {
+			'poll_notifs_sink: loop {
+				// Before we poll the notifications sink receiver, check that all the notification
+				// channels are ready to send a message.
+				for (out_handler, _) in &mut self.out_handlers {
+					match out_handler.poll_ready(cx) {
+						Poll::Ready(_) => {},
+						Poll::Pending => break 'poll_notifs_sink,
+					}
+				}
+
+				let message = match notifications_sink_rx.poll_next_unpin(cx) {
+					Poll::Ready(Some(msg)) => msg,
+					Poll::Ready(None) | Poll::Pending => break,
+				};
+
+				match message {
+					NotificationsSinkMessage::Legacy { message } => {
+						self.legacy.inject_event(LegacyProtoHandlerIn::SendCustomMessage {
+							message
+						});
+					}
+					NotificationsSinkMessage::Notification {
+						protocol_name,
+						encoded_fallback_message,
+						message
+					} => {
+						for (handler, _) in &mut self.out_handlers {
+							if handler.protocol_name() != &protocol_name[..] {
+								continue;
+							}
+
+							if handler.is_open() {
+								handler.send_or_discard(message);
+								continue 'poll_notifs_sink;
+							}
```

Comment:
I can't merge these two, the `continue` refers to different branches.

Reply:
Sure, but I don't see the problem. This
```rust
for (handler, _) in &mut self.out_handlers {
    if handler.protocol_name() != &protocol_name[..] {
        continue;
    }

    if handler.is_open() {
        handler.send_or_discard(message);
        continue 'poll_notifs_sink;
    }
}
```
looks to me equivalent to
```rust
for (handler, _) in &mut self.out_handlers {
    if handler.protocol_name() == &protocol_name[..] && handler.is_open() {
        handler.send_or_discard(message);
        continue 'poll_notifs_sink;
    }
}
```
Both loops jump to `'poll_notifs_sink` as soon as an open handler with a matching protocol name is found, otherwise the loops runs to completion. What difference am I missing?

Reply:
Ah right, I now realize that `continue 'poll_notifs_sink;` should be outside of the `if`.

Reply:
Makes sense to avoid unnecessary iterations. Feel free to resolve this point at your convenience.